### ISSUE TITLE
Increased in binary size on 4.1.0

### ISSUE THREAD

**Issue by jibraniqbal-careem** [Adyen/adyen-android#465]
**Describe the bug**
Increase in binary size

**To Reproduce**
we have updated the library from com.adyen.threeds:3.8.3 to 4.0.0, it increased our binary size from 300kb from older, and on 4.1.0 it's 1.7mb increased, which looks like a big increase from jumping it 33.3 to 33.6 and these internal dependecies my might be culprit from old version to new, we only use 3ds2 component, is there a strip down library for it?
 
New Dependencies
com.adyen.checkout:checkout-core:4.0.0
com.adyen.checkout:components-core:4.0.0
com.adyen.checkout:redirect:4.0.0
com.adyen.checkout:ui-core:4.0.0

Removed Dependencies
android.arch.lifecycle:extensions:1.1.1 -> androidx.lifecycle:lifecycle-extensions:2.2.0
com.adyen.checkout:base-v3:3.8.3
com.adyen.checkout:core-v3:3.8.3

**Expected behavior**
Shouldn't increase binary size by this margin and as rewrite happened on sdk, it progress to decrease the size of it maybe.

**Platform**
 - Android

**Comment by caiofaustino** [comment]
hey there,

So the what happened there is there was a change on the feature set for that library, our 3DS2 component can now also handle a 3DS1 fallback which is a redirect, so now it also depends on the Redirect component. That's probably where the size increase comes from, as redirect also depends on `androidx.browser:browser`.

If the library size is really a big concern for you, and you are sure you do not need to fallback to 3DS1, you can manually exclude the `com.adyen.checkout:redirect:4.0.0` and `com.adyen.checkout:ui-core:4.0.0` dependencis on gradle. Or you can change to the pure 3DS2 SDK.

**Comment by jibraniqbal-careem** [comment]
thanks, i will check, can you provide link how we can now implement 3DS1 using this library, we might want to migrate to that, it would be awesome if there is a function that takes 3DS1 params like bankUrl, PAReq and MD.

**Comment by jibraniqbal-careem** [comment]
excluding 'com.adyen.checkout:redirect:4.0.0' leads to crash on 3ds2

E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.careem.careempay, PID: 14629
    java.lang.NoClassDefFoundError: Failed resolution of: Lcom/adyen/checkout/redirect/RedirectDelegate;
        at com.adyen.checkout.adyen3ds2.Adyen3DS2ComponentProvider.get(Adyen3DS2ComponentProvider.kt:31)
        at com.adyen.checkout.adyen3ds2.Adyen3DS2ComponentProvider.get(Adyen3DS2ComponentProvider.kt:23)
        at com.careem.pay.purchase.widgets.paymentmethods.ThreeDSHandler.launchThreeDS2(ThreeDSHandler.kt:34)
        at com.careem.pay.purchase.widgets.paymentmethods.ThreeDSHandler.launchThreeDS(ThreeDSHandler.kt:26)
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.showOTPRequired(PayPaymentWidget.kt:201)
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.onPaymentStateChanged(PayPaymentWidget.kt:174)
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.setUpPaymentView$lambda-6(PayPaymentWidget.kt:150)
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.$r8$lambda$zQD6VKMlLywaZ6su3c1MhZqytYI(Unknown Source:0)
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget$$ExternalSyntheticLambda5.onChanged(Unknown Source:4)
        at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133)
        at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151)
        at androidx.lifecycle.LiveData.setValue(LiveData.java:309)
        at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
        at androidx.lifecycle.LiveData$1.run(LiveData.java:93)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "com.adyen.checkout.redirect.RedirectDelegate" on path: DexPathList[[zip file "/data/app/com.careem.careempay-228bCp4JdARt3kgQ-kWGcQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.careem.careempay-228bCp4JdARt3kgQ-kWGcQ==/lib/arm64, /data/app/com.careem.careempay-228bCp4JdARt3kgQ-kWGcQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:196)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at com.adyen.checkout.adyen3ds2.Adyen3DS2ComponentProvider.get(Adyen3DS2ComponentProvider.kt:31) 
        at com.adyen.checkout.adyen3ds2.Adyen3DS2ComponentProvider.get(Adyen3DS2ComponentProvider.kt:23) 
        at com.careem.pay.purchase.widgets.paymentmethods.ThreeDSHandler.launchThreeDS2(ThreeDSHandler.kt:34) 
        at com.careem.pay.purchase.widgets.paymentmethods.ThreeDSHandler.launchThreeDS(ThreeDSHandler.kt:26) 
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.showOTPRequired(PayPaymentWidget.kt:201) 
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.onPaymentStateChanged(PayPaymentWidget.kt:174) 
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.setUpPaymentView$lambda-6(PayPaymentWidget.kt:150) 
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget.$r8$lambda$zQD6VKMlLywaZ6su3c1MhZqytYI(Unknown Source:0) 
        at com.careem.pay.purchase.widgets.payment.PayPaymentWidget$$ExternalSyntheticLambda5.onChanged(Unknown Source:4) 
        at androidx.lifecycle.LiveData.considerNotify(LiveData.java:133) 
        at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:151) 
        at androidx.lifecycle.LiveData.setValue(LiveData.java:309) 
        at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50) 
        at androidx.lifecycle.LiveData$1.run(LiveData.java:93) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)

**Comment by jibraniqbal-careem** [comment]
> Yeah sorry I thought removing the dependency could work, but the code is not expecting that.
> 
> The action to automatically handle 3DS1 redirect is available in API v67. It doesn't take the 3DS1 params directly, just the redirect Action.
> 
> Is the binary size increase a big concern for you? Maybe try removing only the `androidx.browser:browser`dependency.


tried excluding browser only 0.1 difference, can we try excluding this org.bouncycastle:bcprov-jdk15to18:1.69?

**Comment by caiofaustino** [comment]
No unfortunately not, bouncy castle is required for 3DS2 to work. I assume you are already applying Proguard/R8 so that should reduce the lib size for bouncy castle considerably.

**Comment by jibraniqbal-careem** [comment]
> No unfortunately not, bouncy castle is required for 3DS2 to work. I assume you are already applying Proguard/R8 so that should reduce the lib size for bouncy castle considerably.

can you provide proguard for that because we didn't do it explicitly, it was not required before, so we don't removed important classes.

**Comment by caiofaustino** [comment]
1.7mb only for the 3DS2 component?

**Comment by jibraniqbal-careem** [comment]
yes only importing this com.adyen.checkout:3ds2:4.1.0, see the dex size.
New version 4.1.0
![image (1)](https://user-images.githubusercontent.com/59829224/131469857-7c43ffe5-7d74-451e-9627-6a9f2b692109.png)
Old version 3.8.3
![image](https://user-images.githubusercontent.com/59829224/131469869-1d2c472e-c6ad-4131-9783-828e3043237d.png)

**Comment by jibraniqbal-careem** [comment]
> 1.7mb only for the 3DS2 component?

https://github.com/Adyen/adyen-android/issues/465#issuecomment-909021619

**Comment by caiofaustino** [comment]
Yeah that's pretty bad, I'll make some tests here and see what we can do.

**Comment by caiofaustino** [comment]
So I just made a quick test here, and the fact that you can even see bouncy castle in the dex analyzer seems that you are not configuring Proguard/R8 to minify and obfuscate your APK. Otherwise it would show up as just letters.

Can you check on your build.gradle how that's configured please? Something like this?
```
buildTypes {
    release {
        minifyEnabled true
        proguardFiles getDefaultProguardFile("proguard-android.txt")
    }
}
```

**Comment by caiofaustino** [comment]
In my test, the raw Bouncy Castle dependency goes from 2,8 Mb in dex size to 236kb after R8,
`org.bouncycastle.asn1.ASN1ApplicationSpecific -> m.a.a.a:`

<img width="1291" alt="Screen Shot 2021-09-01 at 10 51 25" src="https://user-images.githubusercontent.com/3094935/131642036-a2e40e5c-ad9e-4013-b505-96c34bae022a.png">
<img width="1282" alt="Screen Shot 2021-09-01 at 10 51 36" src="https://user-images.githubusercontent.com/3094935/131642040-d18424d3-b987-4491-a325-0be9988ffca5.png">

**Comment by jibraniqbal-careem** [comment]
It works with proguard thanks, great support as always.

**Comment by jibraniqbal-careem** [comment]
Hey guys looks like bouncy-castle is included again in 4.1.0 which increase the binary size i saw in configuration.txt can this be a culprit Crash InvalidInputException #445 

-keep class org.bouncycastle.** {*;}
-dontwarn org.bouncycastle.**

**Comment by caiofaustino** [comment]
hey @jibraniqbal-careem 

yeah the issue on #445 which we believe is the same as [this](https://github.com/Adyen/adyen-3ds2-android/issues/33) was caused by having 2 versions of Bouncy Castle in the build. That's why we made the change to include it as a gradle dependency instead of embedded jar, that way gradle will see that the conflict exists and will resolve it to the latest one available.

I don't think the increase in binary size would come from this change like I mentioned because it gets shrunk by Proguard/R8 in the end.
